### PR TITLE
Commit handle is stored in query state machine

### PR DESCRIPTION
**Pull request by lliiun-z**
After the metastore operations are committed through in memory transaction manager;
the query state is updated in query state machine, which will trigger the rest of
the operations, like logging, etc.

Test plan 
1.  We executed the verifier test and there are no relevant errors found.

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

**Comment by lliiun-z** [comment]
The test failures are irrelevant (from presto-docs).

File: presto-main/src/main/java/com/facebook/presto/transaction/InMemoryTransactionManager.java [hunk 1]
```diff
@@ -568,11 +568,13 @@ public synchronized void checkConnectorWrite(ConnectorId connectorId)
...
             ConnectorTransactionMetadata writeConnector = connectorIdToMetadata.get(writeConnectorId);
             Supplier<ListenableFuture> commitFunctionNamespaceTransactions = () -> functionNamespaceFuture;
-            ListenableFuture<?> commitFuture = Futures.transformAsync(finishingExecutor.submit(writeConnector::commit), ignored -> commitFunctionNamespaceTransactions.get(), directExecutor());
-            ListenableFuture<?> readOnlyCommitFuture = Futures.transformAsync(commitFuture, ignored -> commitReadOnlyConnectors.get(), directExecutor());
-            addExceptionCallback(readOnlyCommitFuture, this::abortInternal);
+            ListenableFuture<?> readOnlyCommitFuture = Futures.transformAsync(
+                    commitFunctionNamespaceTransactions.get(), ignored -> commitReadOnlyConnectors.get(), directExecutor());
+            ListenableFuture<?> writeCommitFuture = Futures.transformAsync(
+                    readOnlyCommitFuture, ignored -> finishingExecutor.submit(writeConnector::commit), directExecutor());
+            addExceptionCallback(writeCommitFuture, this::abortInternal);
```
Comment (NikhilCollooru) [inline presto-main/src/main/java/com/facebook/presto/transaction/InMemoryTransactionManager.java:575]: looks like the commit order is changed here. can you briefly explain this change.

File: presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java [hunk 2]
```diff
@@ -525,6 +529,12 @@ public void setOutput(Optional<Output> output)
         this.output.set(output);
     }
 
+    public void SetConnectorCommitResult(ConnectorCommitHandle commitHandle)
+    {
+        requireNonNull(commitHandle, "commitHandle is null");
+        this.commitHandle.set(commitHandle);
+    }
+
```
[the remark this answers is not in the document]
Comment (lliiun-z) [reply presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java:537]: I should remove it.

File: presto-main/src/main/java/com/facebook/presto/transaction/InMemoryTransactionManager.java (hunk 1 above)
Comment (lliiun-z) [reply presto-main/src/main/java/com/facebook/presto/transaction/InMemoryTransactionManager.java:575]: Previously, the commit result is not returned; therefore, the order of the commits are not important. Here we need to return the commit result for further handling, the order matters. Two cases: for read query commit, the order is the same as previous. For write query commits, write commits should be the last one in order to return the result (e.g., create a table, or add partitions). Previously the write commit results are swallowed before the read commits.

File: presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java [hunk 3]
```diff
@@ -733,13 +744,18 @@ public void onFailure(Throwable throwable)
...
         }
         else {
-            transitionToFinished();
+            transitionToFinished(null);
         }
         return true;
     }
 
-    private void transitionToFinished()
+    public void transitionToFinished(Object result)
```
[the remark this answers is not in the document]
Comment (lliiun-z) [reply presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java:752]: Yeah, I changed it in the middle of coding; need to remove it.

File: presto-main/src/main/java/com/facebook/presto/transaction/InMemoryTransactionManager.java (hunk 1 above)
Comment (NikhilCollooru) [reply presto-main/src/main/java/com/facebook/presto/transaction/InMemoryTransactionManager.java:577]: Can we please add a comment explaining the logic you mentioned above.

Comment (NikhilCollooru) [reply presto-main/src/main/java/com/facebook/presto/transaction/InMemoryTransactionManager.java:577]: Can we please add a comment explaining the logic you mentioned above.

**Comment by NikhilCollooru** [comment]
1. Are there any existing tests that are testing this code path ? If not can we add some tests ?
2. Please update the PR description accordingly.

File: presto-main/src/main/java/com/facebook/presto/transaction/InMemoryTransactionManager.java (hunk 1 above)
Comment (lliiun-z) [reply presto-main/src/main/java/com/facebook/presto/transaction/InMemoryTransactionManager.java:577]: I have added a simple version. Do you think if it is enough?

**Review by NikhilCollooru — APPROVED** [review]

File: presto-main/src/main/java/com/facebook/presto/transaction/InMemoryTransactionManager.java [hunk 4]
```diff
@@ -566,13 +567,16 @@ public synchronized void checkConnectorWrite(ConnectorId connectorId)
...
+            // for transactions with read and write, we only return the commit handle for write query.
             ConnectorTransactionMetadata writeConnector = connectorIdToMetadata.get(writeConnectorId);
             Supplier<ListenableFuture> commitFunctionNamespaceTransactions = () -> functionNamespaceFuture;
-            ListenableFuture<?> commitFuture = Futures.transformAsync(finishingExecutor.submit(writeConnector::commit), ignored -> commitFunctionNamespaceTransactions.get(), directExecutor());
-            ListenableFuture<?> readOnlyCommitFuture = Futures.transformAsync(commitFuture, ignored -> commitReadOnlyConnectors.get(), directExecutor());
-            addExceptionCallback(readOnlyCommitFuture, this::abortInternal);
+            ListenableFuture<?> readOnlyCommitFuture = Futures.transformAsync(
+                    commitFunctionNamespaceTransactions.get(), ignored -> commitReadOnlyConnectors.get(), directExecutor());
+            ListenableFuture<?> writeCommitFuture = Futures.transformAsync(
+                    readOnlyCommitFuture, ignored -> finishingExecutor.submit(writeConnector::commit), directExecutor());
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/transaction/InMemoryTransactionManager.java:576]: Let's put them into the same line:
```java
ListenableFuture<?> readOnlyCommitFuture = Futures.transformAsync(commitFunctionNamespaceTransactions.get(), ignored -> commitReadOnlyConnectors.get(), directExecutor());
ListenableFuture<?> writeCommitFuture = Futures.transformAsync(readOnlyCommitFuture, ignored -> finishingExecutor.submit(writeConnector::commit), directExecutor());

File: presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java [hunk 5]
```diff
@@ -715,14 +719,15 @@ public boolean transitionToFinishing()
...
         Optional<TransactionInfo> transaction = session.getTransactionId()
                 .flatMap(transactionManager::getOptionalTransactionInfo);
+
         if (transaction.isPresent() && transaction.get().isAutoCommitContext()) {
             ListenableFuture<?> commitFuture = transactionManager.asyncCommit(transaction.get().getTransactionId());
             Futures.addCallback(commitFuture, new FutureCallback<Object>()
             {
                 @Override
-                public void onSuccess(@Nullable Object result)
+                public void onSuccess(@Nullable Object commitHandle)
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java:728]: keep the name `result`

File: presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java [hunk 6]
```diff
@@ -733,13 +738,17 @@ public void onFailure(Throwable throwable)
...
         return true;
     }
 
-    private void transitionToFinished()
+    private void transitionToFinished(Object commitHandle)
     {
+        if (commitHandle instanceof ConnectorCommitHandle) {
+            this.commitHandle.set((ConnectorCommitHandle) commitHandle);
+        }
+
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java:751]: Do not change this; check my other comments

File: presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java [hunk 7]
```diff
@@ -733,13 +738,17 @@ public void onFailure(Throwable throwable)
             }, directExecutor());
         }
         else {
-            transitionToFinished();
+            transitionToFinished(null);
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java:741]: we don't pass nulls; it's too dangerous. It's like we pass nullptr in C++

File: presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java [hunk 8]
```diff
@@ -715,14 +719,15 @@ public boolean transitionToFinishing()
...
+
         if (transaction.isPresent() && transaction.get().isAutoCommitContext()) {
             ListenableFuture<?> commitFuture = transactionManager.asyncCommit(transaction.get().getTransactionId());
             Futures.addCallback(commitFuture, new FutureCallback<Object>()
             {
                 @Override
-                public void onSuccess(@Nullable Object result)
+                public void onSuccess(@Nullable Object commitHandle)
                 {
-                    transitionToFinished();
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java:725]:
```java
if (result instanceof ConnectorCommitHandle) {
    this.commitHandle.set((ConnectorCommitHandle) commitHandle);
}

**Review by highker — APPROVED** [review]

File: presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java (hunk 7 above)
Comment (lliiun-z) [reply presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java:741]: Then what to pass then? How about a magic number?

Comment (lliiun-z) [reply presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java]: Use Optional.empty() as the parameter.

File: presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java (hunk 8 above)
Comment (highker) [reply presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java:725]: I mean let's move the code logic here and do not change the interface for `transitionToFinished`.

File: presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java [hunk 9]
```diff
@@ -733,13 +738,17 @@ public void onFailure(Throwable throwable)
...
         }
         return true;
     }
 
-    private void transitionToFinished()
+    private void transitionToFinished(Object result)
     {
+        if (result instanceof ConnectorCommitHandle) {
+            this.commitHandle.set((ConnectorCommitHandle) result);
+        }
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java:750]: Let's keep the function untouched.

**Pull request merged**